Lib/traceback.py

```diff
@@ -622,15 +630,32 @@ def format(self, *, chain=True):
         The message indicating which exception occurred is always the last
         string in the output.
         """
-        if chain:
-            if self.__cause__ is not None:
-                yield from self.__cause__.format(chain=chain)
-                yield _cause_message
-            elif (self.__context__ is not None and
-                not self.__suppress_context__):
-                yield from self.__context__.format(chain=chain)
-                yield _context_message
-        if self.stack:
-            yield 'Traceback (most recent call last):\n'
-            yield from self.stack.format()
-        yield from self.format_exception_only()
+
+        stack = []
+        exc = self
+        while exc:
+            if chain:
+                if exc.__cause__:
+                    chained_msg = _cause_message
+                    chained_exc = exc.__cause__
+                elif exc.__context__ and not exc.__suppress_context__:
+                    chained_msg = _context_message
+                    chained_exc = exc.__context__
+                else:
+                    chained_msg = None
+                    chained_exc = None
+
+                stack.append((chained_msg, exc))
+                exc = chained_exc
+            else:
+                stack.append((None, exc))
+                exc = None
+
+        while stack:
+            msg, exc = stack.pop()
```

Comment:
A faster way might be to use `for msg, exc in reversed(stack):` (this may also give you inspiration for what to name the variable. :-)